src/Microsoft.ML.Core/Data/Schema.cs

```diff
@@ -0,0 +1,346 @@
+            Contracts.CheckNonEmpty(name, nameof(name));
+            if (_nameMap.TryGetValue(name, out int col))
+                return _columns[col];
+            return null;
+        }
+
+        /// <summary>
+        /// This class describes one column in the schema.
+        /// </summary>
+        public sealed class Column
+        {
+            /// <summary>
+            /// The name of the column.
+            /// </summary>
+            public string Name { get; }
+
+            /// <summary>
+            /// The type of the column.
+            /// </summary>
+            public ColumnType Type { get; }
+
+            /// <summary>
+            /// The metadata of the column.
+            /// </summary>
+            public MetadataRow Metadata { get; }
+
+            public Column(string name, ColumnType type, MetadataRow metadata)
+            {
+                Contracts.CheckNonEmpty(name, nameof(name));
+                Contracts.CheckValue(type, nameof(type));
+                Contracts.CheckValueOrNull(metadata);
+
+                Name = name;
+                Type = type;
+                Metadata = metadata;
+            }
+        }
+
+        /// <summary>
+        /// The metadata of one <see cref="Column"/>.
+        /// </summary>
+        public sealed class MetadataRow
+        {
+            private readonly (Column column, Delegate getter)[] _values;
+
+            /// <summary>
+            /// The schema of the metadata row. It is different from the schema that the column belongs to.
+            /// </summary>
+            public Schema Schema { get; }
+
+            /// <summary>
+            /// Create a metadata row by supplying the schema columns and the getter delegates for all the values.
+            /// </summary>
+            public MetadataRow(IEnumerable<(Column column, Delegate getter)> values)
+            {
+                Contracts.CheckValue(values, nameof(values));
+                // Check all getters.
+                foreach (var (column, getter) in values)
+                {
+                    Contracts.CheckValue(column, nameof(column));
+                    Contracts.CheckValue(getter, nameof(getter));
+                    Utils.MarshalActionInvoke(CheckGetter<int>, column.Type.RawType, getter);
+                }
+                _values = values.ToArray();
+                Schema = new Schema(_values.Select(x => x.column));
+            }
+
+            private void CheckGetter<TValue>(Delegate getter)
+            {
+                var typedGetter = getter as ValueGetter<TValue>;
+                if (typedGetter == null)
+                    throw Contracts.ExceptParam(nameof(getter), $"Getter of type '{typeof(TValue)}' expected, but {getter.GetType()} found");
+            }
+
+            /// <summary>
+            /// Get a getter delegate for one value of the metadata row.
+            /// </summary>
+            public ValueGetter<TValue> GetGetter<TValue>(int col)
```

Comment:
>GetGetter [](start = 39, length = 9)

Does this method (and the next one) have to be public?

Reply:
GetGetter does. GetValue by col didn't

---
In reply to: [224846761](https://github.com/dotnet/machinelearning/pull/1167#discussion_r224846761) [](ancestors = 224846761)